contracts/token/ERC20/extensions/ERC4626.sol

```diff
@@ -0,0 +1,187 @@
+        require(shares <= maxMint(receiver), "ERC4626: mint more than max");
+
+        address caller = _msgSender();
+        uint256 assets = previewMint(shares);
+
+        // if _asset is ERC777, transferFrom can call reenter BEFORE the transfer happens through
+        // the tokensToSend hook, so we need to transfer before we mint to keep the invariants.
+        SafeERC20.safeTransferFrom(_asset, caller, address(this), assets);
+        _mint(receiver, shares);
+
+        emit Deposit(caller, receiver, assets, shares);
+
+        return assets;
+    }
+
+    /** @dev See {IERC4262-withdraw} */
+    function withdraw(
+        uint256 assets,
+        address receiver,
+        address owner
+    ) public virtual override returns (uint256) {
+        require(assets <= maxWithdraw(owner), "ERC4626: withdraw more than max");
+
+        address caller = _msgSender();
+        uint256 shares = previewWithdraw(assets);
+
+        if (caller != owner) {
+            _spendAllowance(owner, caller, shares);
+        }
+
+        // if _asset is ERC777, transfer can call reenter AFTER the transfer happens through
+        // the tokensReceived hook, so we need to transfer after we burn to keep the invariants.
+        _burn(owner, shares);
+        SafeERC20.safeTransfer(_asset, receiver, assets);
+
+        emit Withdraw(caller, receiver, owner, assets, shares);
+
+        return shares;
+    }
+
+    /** @dev See {IERC4262-redeem} */
+    function redeem(
+        uint256 shares,
+        address receiver,
+        address owner
+    ) public virtual override returns (uint256) {
+        require(shares <= maxRedeem(owner), "ERC4626: redeem more than max");
+
+        address caller = _msgSender();
+        uint256 assets = previewRedeem(shares);
+
+        if (caller != owner) {
+            _spendAllowance(owner, caller, shares);
+        }
+
+        // if _asset is ERC777, transfer can call reenter AFTER the transfer happens through
+        // the tokensReceived hook, so we need to transfer after we burn to keep the invariants.
+        _burn(owner, shares);
+        SafeERC20.safeTransfer(_asset, receiver, assets);
+
+        emit Withdraw(caller, receiver, owner, assets, shares);
+
+        return assets;
+    }
+
+    /**
+     * @dev Internal convertion function (from assets to shares) with support for rounding direction
+     *
+     * Will revert if assets > 0, totalSupply > 0 and totalAssets = 0. That corresponds to a case where any asset
+     * would represent an infinite amout of shares.
+     */
+    function _convertToShares(uint256 assets, Math.Rounding direction) internal view virtual returns (uint256 shares) {
+        uint256 supply = totalSupply();
+        return
+            (assets == 0 || supply == 0)
+                ? assets.mulDiv(10**decimals(), 10**_asset.decimals(), direction)
+                : assets.mulDiv(supply, totalAssets(), direction);
```

Comment:
I see that when the vault is empty of assets or shares the assets/shares ratio is set to 1:1. After that `totalAssets()` is used to calculate the ratio. I was considering the case where some amount of asset is sent to the contract (direct transfer) before the first `_mint()`. The first deposit would get 1:1, but subsequent deposits would get significantly less because of the existing initial balance. I understand that this functions will be more likely overridden by the integrator with the specific minting strategy, so I'm not sure if it should be fixed here, but I thought it was worth mentioning it

Reply:
I think this is something that we document in the tests

If a vault has some assets but no shares, then the value of shares is infinite, and the "logic" would mean that every share as an infinite value, and any assets put in by the user would give them 0 shares ... that is obviously not good.

So yes, the first depositor gets all the previous assets for free ... but it would also have got them for free if they had been transferred to the vault just after the deposit.

Also, I'm not sure how integrators are going to override this. I'm expecting that quite a few implementations don't override it at all.

Reply:
About integrators, I just imagined a Vault taking into account special addresses, or bonus periods, or something along those lines. Thanks for the answer!

Reply:
We shouldn't expect this to be overriden, the contract should have reasonable behavior out of the box.

>So yes, the first depositor gets all the previous assets for free ... but it would also have got them for free if they had been transferred to the vault just after the deposit.

The fact that these things commute makes me think the current behavior is reasonable.